File: benchmark/task-queue/tasks.go
package main

import "fmt"

const (
	StatusPending   = 0
	StatusRunning   = 1
	StatusCompleted = 2
	StatusFailed    = 3
)

const (
	PriorityLow    = 0
	PriorityMedium = 1
	PriorityHigh   = 2
)

type Task struct {
	id       int
	priority int
	status   int
	duration int
	retries  int
}

type TaskQueue struct {
	tasks     []Task
	completed int
	failed    int
	totalTime int
}

func newTaskQueue() *TaskQueue {
	return &TaskQueue{
		tasks:     make([]Task, 0),
		completed: 0,
		failed:    0,
		totalTime: 0,
	}
}

func (q *TaskQueue) addTask(id, priority, duration int) {
	task := Task{
		id:       id,
		priority: priority,
		status:   StatusPending,
		duration: duration,
		retries:  0,
	}
	q.tasks = append(q.tasks, task)
}

func (q *TaskQueue) findHighestPriority() int {
	bestIdx := -1
	bestPriority := -1
	for i := 0; i < len(q.tasks); i++ {
		if q.tasks[i].status == StatusPending {
			if q.tasks[i].priority > bestPriority {
				bestPriority = q.tasks[i].priority
				bestIdx = i
			}
		}
	}
	return bestIdx
}

func (q *TaskQueue) processTask(idx int, currentTime int) bool {
	if idx < 0 || idx >= len(q.tasks) {
		return false
	}
	
	task := &q.tasks[idx]
	task.status = StatusRunning
	
	// Simulate processing: tasks with odd duration "fail" on first try
	shouldFail := task.duration%2 == 1 && task.retries == 0
	
	if shouldFail {
		task.retries++
		task.status = StatusPending // Retry
		return false
	}
	
	// Check timeout (tasks > 100 duration fail after 3 retries)
	if task.duration > 100 && task.retries >= 3 {
		task.status = StatusFailed
		q.failed++
		return false
	}
	
	// Success
	task.status = StatusCompleted
	q.completed++
	q.totalTime += task.duration
	return true
}

func (q *TaskQueue) runUntilEmpty() int {
	iterations := 0
	for {
		idx := q.findHighestPriority()
		if idx < 0 {
			break
		}
		q.processTask(idx, iterations)
		iterations++
		if iterations > 1000000 {
			break // Safety limit
		}
	}
	return iterations
}

func (q *TaskQueue) getStats() (int, int, int) {
	return q.completed, q.failed, q.totalTime
}

func main() {
	totalCompleted := 0
	totalFailed := 0
	totalTime := 0
	
	// Run multiple rounds
	rounds := 1000
	tasksPerRound := 100
	
	for round := 0; round < rounds; round++ {
		q := newTaskQueue()
		
		// Add tasks with varying priorities and durations
		for i := 0; i < tasksPerRound; i++ {
			priority := i % 3
			duration := (i * 7 + round) % 150
			q.addTask(round*tasksPerRound+i, priority, duration)
		}
		
		q.runUntilEmpty()
		
		completed, failed, time := q.getStats()
		totalCompleted += completed
		totalFailed += failed
		totalTime += time
	}
	
	fmt.Println(totalCompleted)
	fmt.Println(totalFailed)
	fmt.Println(totalTime)
}

File: benchmark/task-queue/tasks.py
STATUS_PENDING = 0
STATUS_RUNNING = 1
STATUS_COMPLETED = 2
STATUS_FAILED = 3

PRIORITY_LOW = 0
PRIORITY_MEDIUM = 1
PRIORITY_HIGH = 2

class Task:
    def __init__(self, id, priority, duration):
        self.id = id
        self.priority = priority
        self.status = STATUS_PENDING
        self.duration = duration
        self.retries = 0

class TaskQueue:
    def __init__(self):
        self.tasks = []
        self.completed = 0
        self.failed = 0
        self.total_time = 0
    
    def add_task(self, id, priority, duration):
        self.tasks.append(Task(id, priority, duration))
    
    def find_highest_priority(self):
        best_idx = -1
        best_priority = -1
        for i, task in enumerate(self.tasks):
            if task.status == STATUS_PENDING:
                if task.priority > best_priority:
                    best_priority = task.priority
                    best_idx = i
        return best_idx
    
    def process_task(self, idx, current_time):
        if idx < 0 or idx >= len(self.tasks):
            return False
        
        task = self.tasks[idx]
        task.status = STATUS_RUNNING
        
        # Simulate processing: tasks with odd duration "fail" on first try
        should_fail = task.duration % 2 == 1 and task.retries == 0
        
        if should_fail:
            task.retries += 1
            task.status = STATUS_PENDING  # Retry
            return False
        
        # Check timeout (tasks > 100 duration fail after 3 retries)
        if task.duration > 100 and task.retries >= 3:
            task.status = STATUS_FAILED
            self.failed += 1
            return False
        
        # Success
        task.status = STATUS_COMPLETED
        self.completed += 1
        self.total_time += task.duration
        return True
    
    def run_until_empty(self):
        iterations = 0
        while True:
            idx = self.find_highest_priority()
            if idx < 0:
                break
            self.process_task(idx, iterations)
            iterations += 1
            if iterations > 1000000:
                break  # Safety limit
        return iterations
    
    def get_stats(self):
        return self.completed, self.failed, self.total_time

def main():
    total_completed = 0
    total_failed = 0
    total_time = 0
    
    rounds = 2000
    tasks_per_round = 150
    
    for round in range(rounds):
        q = TaskQueue()
        
        for i in range(tasks_per_round):
            priority = i % 3
            duration = (i * 7 + round) % 150
            q.add_task(round * tasks_per_round + i, priority, duration)
        
        q.run_until_empty()
        
        completed, failed, time = q.get_stats()
        total_completed += completed
        total_failed += failed
        total_time += time
    
    print(total_completed)
    print(total_failed)
    print(total_time)

if __name__ == "__main__":
    main()
